### PR TITLE
Implement the `anchor-handler` contract

### PR DESCRIPTION
**Pull request by duguorong009**
Summary of changes
Changes introduced in this pull request

Implement the `anchor-handler` contract
   - Implement `handler` common utils
   - Implement `execute_proposal` entry

Reference issue to close (if applicable)
Closes

### CODE REVIEW

File: contracts/anchor-handler/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+
+        /* ---  Anchor-handler specific execution entries --- */
+        // @notice Proposal execution should be initiated when a proposal is finalized in the Bridge contract.
+        // by a relayer on the deposit's destination chain.
+        //
+        // Params:
+        //   "resource_id": ResourceID corresponding to a particular set of Anchors
+        //   "data": Consists of {sourceChainID}, {leafIndex}, {merkleRoot} all padded to 32 bytes.
+        //
+        // Data passed into the function should be constructed as follows:
```
Comment (drewstone) [inline contracts/anchor-handler/src/contract.rs:90]: On Substrate we do encode the `call` of the function being executed into the proposal data, so I would be curious if you could provide details on doing that here as well. It might make sense or not, since if we do use the encoded function call directly we'd need to maintain yet another set of proposals for CosmWasm.

File: contracts/anchor-handler/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    let anchor_addr = read_contract_addr(deps.storage, resource_id)?;
+    if !read_whitelist(deps.storage, anchor_addr)? {
+        return Err(ContractError::Std(StdError::GenericErr {
+            msg: "provided tokenAddress is not whitelisted".to_string(),
+        }));
+    }
+
+    // Execute the proposal according to function signature
+    let set_handler_sig = bytes4_encoder(
+        &Keccak256::hash(b"setHandler(address,uint32)").map_err(|_| ContractError::DecodeError)?,
```
Comment (drewstone) [inline contracts/anchor-handler/src/contract.rs:165]: Can we "mock" the function signatures as they are on CosmWasm? For example `uint32` would be `u32`, etc.

**Review by drewstone — CHANGES_REQUESTED** [review]
This looks good and is in line with the Solidity protocol. It looks like there are lints to be fixed here. I would be curious what you think is best regarding following the Solidity protocol or not, for example, by encoding the execution function in the message directly or doing it the way we currently are. We can consider keeping the Solidity function signatures as well, which may even provide us more consistency in the data we're sending. Good work on this.

File: contracts/anchor-handler/src/contract.rs (hunk 1 above)
Comment (duguorong009) [reply contracts/anchor-handler/src/contract.rs]: I think it is okay to follow the `solidity` implementation.
The reason is that 
  - it is good for consistency of system
  - In `substrate`, it also does parsing proposal data. which means no difference in essence.

File: contracts/anchor-handler/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+            msg: "provided tokenAddress is not whitelisted".to_string(),
+        }));
+    }
+
+    // Execute the proposal according to function signature
+    let set_handler_sig = bytes4_encoder(
+        &Keccak256::hash(b"SetHandler(String,u32)").map_err(|_| ContractError::DecodeError)?,
+    );
+    let update_edge_sig = bytes4_encoder(
+        &Keccak256::hash(b"UpdateEdge(u64,[u8;32],u32,[u8;32])")
```
Comment (drewstone) [inline contracts/anchor-handler/src/contract.rs:163]: These should match the function itself. It should be the exact snake_case value not the camelCase proposal.

**Review by drewstone — CHANGES_REQUESTED** [review]
Function signatures look wrong.

File: contracts/anchor-handler/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            msg: "Invalid resource id".to_string(),
+        }));
+    }
+    let anchor_addr = read_contract_addr(deps.storage, resource_id)?;
+    if !read_whitelist(deps.storage, anchor_addr.clone())? {
+        return Err(ContractError::Std(StdError::GenericErr {
+            msg: "provided tokenAddress is not whitelisted".to_string(),
+        }));
+    }
+    let proposal_sig = bytes4_encoder(
```
[the remark this answers is not in the document]
Comment (drewstone) [reply contracts/anchor-handler/src/contract.rs:156]: Unless you think this does improve security (but I don't think it does).

File: contracts/anchor-handler/src/tests.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        resource_id: RESOURCE_ID,
+        contract_addr: "anchor-contract".to_string(),
+    };
+    let _res = execute(deps.as_mut(), mock_env(), info, set_resource_msg).unwrap();
+
+    // Try to "execute_proposal"
+    let info = mock_info(BRIDGE, &[]);
+
+    let proposal = LinkableAnchorExecMsg::SetHandler {
+        handler: "new-handler".to_string(),
```
Comment (drewstone) [inline contracts/anchor-handler/src/tests.rs:129]: This should be a CW contract address, can we use a random address so it looks more proper?

**Review by drewstone — CHANGES_REQUESTED** [review]
We should improve the tests and also add tests where the anchor handler is used to update the edge of an anchor contract.

File: contracts/anchor-handler/src/contract.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (duguorong009) [reply contracts/anchor-handler/src/contract.rs]: Remove the `signature` check as recommended. [82a4840](https://github.com/webb-tools/protocol-cosmwasm/pull/43/commits/82a48401d1a386c872937ba502afe11b1cee1d1f)
